### Notebook to build a specials measure

A notebook to help build a specials measure.

- [Imports](#IMPORTS)
- [Availability restriction column analysis](#f1)
- [Explore availability restriction column data](#f2)
- [Restricted Availability - 2](#f2-2)

<a id='IMPORTS'></a>
### Imports
Import libraries required for analysis

In [1]:
#import libraries required for analysis
from ebmdatalab import bq
import os
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt

#set to display all rows in data
#pd.set_option('display.max_rows', None)
pd.option_context('display.float_format', '{:,.2f}'.format)

# Set display options for full width
pd.set_option('display.max_colwidth', None)  # Show full column content
pd.set_option('display.width', None)        # Adjust display width to the notebook's width

<a id='f1'></a>
### Availability restriction column analysis
AMP table contains a column 'avail_restrict'. This uses a code to detail availability of products.
The table 'availabilityrestriction' details the meaning behind these codes.

In [2]:
sql = '''
SELECT * FROM dmd.availabilityrestriction
'''
avail_rest_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_rest.csv'))
avail_rest_DF = avail_rest_DF.fillna('None')
avail_rest_DF

cd                             descr
0   1                              None
1   2           Restricted Availability
2   3         Individual Patient Supply
3   4                          Imported
4   5                    Clinical Trial
5   6                           Special
6   7                            Extemp
7   9                     Not available
8  10  For Patient Safety Consideration

We can count the number of AMPs associated with each avail_restrict value.

In [3]:
sql = '''
WITH amp_ar_count AS (
    SELECT avail_restrict, COUNT(*) AS row_count
    FROM `ebmdatalab.dmd.amp` AS amp
    GROUP BY avail_restrict
)
SELECT
    ar.cd,
    ar.descr, 
    SUM(COALESCE(amp_ar_count.row_count, 0)) AS row_count
FROM `dmd.availabilityrestriction` AS ar
LEFT JOIN amp_ar_count
ON ar.cd = amp_ar_count.avail_restrict
GROUP BY ar.descr, ar.cd
ORDER BY ar.cd;
'''
avail_restrict_df = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_restrict_df.csv'))
avail_restrict_df

cd                             descr  row_count
0   1                               NaN      89895
1   2           Restricted Availability         56
2   3         Individual Patient Supply         24
3   4                          Imported        659
4   5                    Clinical Trial          1
5   6                           Special       7281
6   7                            Extemp          0
7   9                     Not available      61299
8  10  For Patient Safety Consideration          4

We can exclude items where avail_restrict = 1 (no restriction) and 9 (not available). 
There are also zero items with avail_restrict = 7 (Extemp).
The remaining values are worth exploring in more detail to better understand.

<a id='f2'></a>
### Explore availability restriction column data
We can extract the rows with avail_restrict values not 1, 7, 9 to explore further what products they contain. We can then try to identify the significance of these products by looking for number of items and spend in the English Prescribing data.

<a id='f2-2'></a>
#### Restricted Availability (avail_restrict 2)

In [4]:
sql = '''
SELECT id, nm, descr, bnf_code 
FROM `ebmdatalab.dmd.amp` 
WHERE avail_restrict = 2
'''
avail_rest_2_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_rest_2.csv'))
avail_rest_2_DF

id  \
0   41729811000001105   
1   42388711000001103   
2   42527511000001100   
3   42702411000001107   
4   43792511000001100   
5   43792811000001102   
6   44145611000001106   
7   42977611000001107   
8   42977211000001105   
9    3830311000001101   
10   3832011000001100   
11   7972911000001104   
12   7974011000001109   
13  15498911000001105   
14  15505711000001105   
15  15506011000001104   
16   8838911000001102   
17   8839511000001103   
18  37687211000001104   
19  37687511000001101   
20  37687811000001103   
21  37688211000001100   
22  37688811000001104   
23  17462311000001107   
24  17462011000001109   
25  24576611000001105   
26  33659811000001109   
27  42174111000001106   
28  33493111000001108   
29  39473011000001103   
30  42645911000001106   
31  40384611000001108   
32  40556911000001102   
33  41178911000001101   
34  41239511000001102   
35  42020711000001107   
36  42098011000001100   
37  42098911000001101   
38  42659911000001106   
39  43111411000001101   
40  43112711000001100   
41  43113211000001101   
42  43471411000001105   
43  41343811000001106   
44  42023311000001102   
45  42985911000001104   
46  43870811000001108   
47  40306411000001101   
48  40483111000001109   
49  40712911000001109   
50  41780111000001104   
51  42092311000001100   
52  42223111000001107   
53  34765811000001105   
54  39367111000001105   
55  41404811000001109   
56   4820311000001102   

                                                                                                                                                             nm  \
0                                                                                                Vyvgart 400mg/20ml concentrate for solution for infusion vials   
1                                                                                                                                         Nexpovio 20mg tablets   
2                                                                                                                               Livmarli 9.5mg/ml oral solution   
3                                                                                                             Vyvgart 1000mg/5.6ml solution for injection vials   
4                                                                                                 Leqembi 200mg/2ml concentrate for solution for infusion vials   
5                                                                                                 Leqembi 500mg/5ml concentrate for solution for infusion vials   
6                    Comirnaty KP.2 adults and adolescents from 12 years COVID-19 mRNA Vaccine 30micrograms/0.3ml dose dispersion for injection multidose vials   
7                                                                                                                                        Clozapine 25mg tablets   
8                                                                                                                                       Clozapine 100mg tablets   
9                                                                                                                                         Clozaril 25mg tablets   
10                                                                                                                                       Clozaril 100mg tablets   
11                                                                                                                                       Denzapine 25mg tablets   
12                                                                                                                                      Denzapine 100mg tablets   
13                                                                                                                                       Denzapine 50mg tablets   
14                                                                                                                         

In [5]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    SELECT DISTINCT bnf_code 
    FROM `ebmdatalab.dmd.amp` 
    WHERE avail_restrict = 2
)
GROUP BY bnf_code, bnf_name
ORDER BY items DESC
'''
avail_rest_2_usage_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_restrict_2_usage.csv'))

avail_rest_2_usage_DF

bnf_code  \
0   0402010C0AAABAB   
1   090607000BBLJA0   
2   0402010C0AAAAAA   
3   0402010C0AAABAB   
4   0402010C0AAAAAA   
5   0402010C0BCABAB   
6   090607000BBLPA0   
7   0402010C0BBABAB   
8   0402010C0BBABAB   
9   1404000AMBBAAAA   
10  0402010C0BCAAAA   
11  1404000U0BCAAAA   
12  0402010C0BBAAAA   
13  0402010C0BBAAAA   
14  0402010C0BCABAB   
15  1404000U0BCAAAA   
16  0402010C0BCACAC   
17  0402010C0BCAAAA   
18  1404000F0BQAAAZ   
19  0402010C0BCACAC   
20  0402010C0BDABAB   
21  0402010C0BCADAD   
22  0402010C0BDABAB   
23  0402010C0BDAAAA   
24  1404000F0BSAABC   
25  0402010C0BCAEAE   
26  0402010C0BDAAAA   
27  0402010C0BCADAD   
28  0402010C0BDADAI   
29  0402010C0BDAEAJ   
30  0402010C0BDAFAK   
31  0402010C0BDACAH   
32  1404000ATBGAJBQ   
33  1404000ATBHADBK   
34  0402010C0BCAEAE   
35  0402010C0BDAGAL   
36  1404000ATBGAGBL   
37  0801050AZBBABAB   
38  1404000ATBHAFBW   
39  0801050AZBBAAAA   
40  0801050AZBBABAB   
41  1404000ATBGAFBF   
42  1404000AWBBAAAA   

                                                        bnf_name  items  \
0                                        Clozapine 100mg tablets  12588   
1                         Healthy Start Children's Vitamin drops   9901   
2                                         Clozapine 25mg tablets   7506   
3                                            Clozapine_Tab 100mg   5808   
4                                             Clozapine_Tab 25mg   3425   
5                                        Denzapine 100mg tablets   2833   
6                          Healthy Start Women's Vitamin tablets   2633   
7                                         Clozaril 100mg tablets   1967   
8                                             Clozaril_Tab 100mg   1840   
9                  Gardasil 9 vacc inj 0.5ml pre-filled syringes   1451   
10                                        Denzapine 25mg tablets   1312   
11                                 Stamaril vacc inj 0.5ml vials   1272   
12                                             Clozaril_Tab 25mg   1102   
13                                         Clozaril 25mg tablets   1038   
14                                           Denzapine_Tab 100mg    687   
15                      Stamaril_Yellow Fever Vac 0.5ml Vl + Dil    427   
16                                        Denzapine 50mg tablets    232   
17                                            Denzapine_Tab 25mg    198   
18              Infanrix Hexa vacc inj 0.5ml pre-filled syringes    132   
19                                            Denzapine_Tab 50mg    132   
20                                         Zaponex 100mg tablets    120   
21                                           Denzapine_Tab 200mg     79   
22                                             Zaponex_Tab 100mg     49   
23                                              Zaponex_Tab 25mg     45   
24                 Vaxelis vaccine inj 0.5ml pre-filled syringes     43   
25                             Denzapine 50mg/ml oral suspension     29   
26                                          Zaponex 25mg tablets     29   
27                                       Denzapine 200mg tablets     22   
28                           Zaponex 25mg orodispersible tablets     21   
29                           Zaponex 50mg orodispersible tablets     15   
30                          Zaponex 100mg orodispersible tablets     15   
31                         Zaponex 12.5mg orodispersible tablets      9   
32     Comirnaty Omicron XBB.1.5 COVID-19 Vacc single dose vials      7   
33         Spikevax XBB.1.5 COVID-19 Vacc 0.1mg/1ml inj md vials      4   
34                                   Denzapine_Oral Susp 50mg/ml      3   
35                          Zaponex 200mg orodispersible tablets      3   
36              Comirnaty Omicron XBB.1.5 COVID-19 Vacc md vials      3   
37                                            Votrient_Tab 400mg      2   
38  Spikevax JN.1 COVID-19 mRNA Vacc 50microg/0.5ml dose inj pfs      1  

**Restricted availability** denotes medications which are only prescribed from specialist services - not relevent to "specials" measure

<a id='f2-3'></a>
#### Individual Patient Supply (avail_restrict 3)

In [6]:
sql = '''
SELECT id, nm, descr, bnf_code 
FROM `ebmdatalab.dmd.amp` 
WHERE avail_restrict = 3
'''
avail_rest_3_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_rest_3.csv'))
avail_rest_3_DF

id  \
0   15514711000001108   
1   35820711000001107   
2   35821111000001100   
3   35821511000001109   
4   39782811000001104   
5   42561311000001109   
6   43998011000001102   
7   43998811000001108   
8   43999211000001102   
9   44013311000001103   
10  38719711000001100   
11   3797411000001100   
12   3681911000001106   
13   3681311000001105   
14   4085511000001105   
15   3840711000001103   
16   3717611000001103   
17  22652611000001109   
18  12848311000001106   
19  20544711000001107   
20  20543611000001101   
21  20543411000001104   
22  20543211000001103   
23  20542911000001100   

                                                                        nm  \
0   Factor XI 1,000unit powder and solvent for solution for infusion vials   
1                           Crysvita 10mg/1ml solution for injection vials   
2                           Crysvita 20mg/1ml solution for injection vials   
3                           Crysvita 30mg/1ml solution for injection vials   
4                                      Vistogard 10g oral granules sachets   
5                         Oxervate 20micrograms/ml eye drops 1ml unit dose   
6    Yorvipath 168micrograms/0.56ml solution for injection pre-filled pens   
7    Yorvipath 294micrograms/0.98ml solution for injection pre-filled pens   
8     Yorvipath 420micrograms/1.4ml solution for injection pre-filled pens   
9                         Dasiglucagon 4mg/1ml solution for infusion vials   
10               Ilomedin 100micrograms/1ml solution for infusion ampoules   
11                                             Sodium Amytal 60mg capsules   
12                                            Sodium Amytal 200mg capsules   
13                                                   Soneryl 100mg tablets   
14                                            Seconal Sodium 50mg capsules   
15                                           Seconal Sodium 100mg capsules   
16                                                   Tuinal 100mg Pulvules   
17                                                    Talidex 25mg tablets   
18                               Desmospray 2.5micrograms/dose nasal spray   
19                                     BioBag 400 10cm x 10cm dressing kit   
20                                      BioBag 300 6cm x 12cm dressing kit   
21                                       BioBag 200 5cm x 6cm dressing kit   
22                                       BioBag 100 4cm x 5cm dressing kit   
23                                      BioBag 50 2.5cm x 4cm dressing kit   

                                                                                                   descr  \
0   Factor XI 1,000unit powder and solvent for solution for infusion vials (Bio Products Laboratory Ltd)   
1                                       Crysvita 10mg/1ml solution for injection vials (Kyowa Kirin Ltd)   
2                                       Crysvita 20mg/1ml solution for injection vials (Kyowa Kirin Ltd)   
3                                       Crysvita 30mg/1ml solution for injection vials (Kyowa Kirin Ltd)   
4                             Vistogard 10g oral granules sachets (Protherics Medicines Development Ltd)   
5                                  Oxervate 20micrograms/ml eye drops 1ml unit dose (Chapper Healthcare)   
6         Yorvipath 168micrograms/0.56ml solution for injection pre-filled pens (Ascendis Pharma UK Ltd)   
7         Yorvipath 294micrograms/0.98ml solution for injection pre-filled pens (Ascendis Pharma UK Ltd)   
8          Yorvipath 420micrograms/1.4ml solution for injection pre-filled pens (Ascendis Pharma UK Ltd)   
9                                        Dasiglucagon 4mg/1ml solution for infusion vials (WEP Clinical)   
10                         Ilomedin 100micrograms/1ml solution for infusion ampoules (Imported (Greece))   
11                                                           Sodium Amytal 60mg capsules (Special Order)   
12                                    

In [7]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    SELECT DISTINCT bnf_code 
    FROM `ebmdatalab.dmd.amp` 
    WHERE avail_restrict = 3
)
GROUP BY bnf_code, bnf_name
ORDER BY items DESC
'''
avail_rest_3_usage_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_restrict_3_usage.csv'))

avail_rest_3_usage_DF

bnf_code                                   bnf_name  items  \
0   0401030E0BBAAAA                Sodium Amytal 60mg capsules   7481   
1   0401030T0BCAAAC                      Tuinal 100mg Pulvules   7313   
2   0401030T0BCAAAC                       Tuinal_Pulvule 100mg   7279   
3   0401030E0BBAAAA                    Sod Amytal_Pulvule 60mg   6063   
4   0401030H0BBAAAB                          Soneryl_Tab 100mg   2650   
5   0401030H0BBAAAB                      Soneryl 100mg tablets   1859   
6   1311070S0BDAEAA         BioBag 50 2.5cm x 4cm dressing kit   1619   
7   0401030T0BBABAB              Seconal Sodium 100mg capsules   1610   
8   0401030T0BBABAB                  Seconal Sod_Pulvule 100mg   1519   
9   1311070S0BDADAA          BioBag 100 4cm x 5cm dressing kit   1395   
10  1311070S0BDACAA          BioBag 200 5cm x 6cm dressing kit   1240   
11  0401030E0BBABAB                   Sod Amytal_Pulvule 200mg    979   
12  1311070S0BDABAA         BioBag 300 6cm x 12cm dressing kit    900   
13  1311070S0BDAAAA        BioBag 400 10cm x 10cm dressing kit    814   
14  1311070S0BDAEAA            BioBag_50 Dress Kit 2.5cm x 4cm    417   
15  0401030T0BBAAAA               Seconal Sodium 50mg capsules    303   
16  0401030E0BBABAB               Sodium Amytal 200mg capsules    302   
17  1311070S0BDACAA             BioBag_200 Dress Kit 5cm x 6cm    289   
18  1311070S0BDADAA             BioBag_100 Dress Kit 4cm x 5cm    287   
19  1311070S0BDABAA            BioBag_300 Dress Kit 6cm x 12cm    197   
20  0401030T0BBAAAA                   Seconal Sod_Pulvule 50mg    196   
21  0605020E0BCACBP  Desmospray 2.5micrograms/dose nasal spray    175   
22  1311070S0BDAAAA           BioBag_400 Dress Kit 10cm x 10cm    169   
23  0605020E0BCACBP            Desmospray_Nsl Spy 2.5mcg (50D)     99   
24  0501100J0BCAAAJ                       Talidex 25mg tablets     19   
25  0501100J0BCAAAJ                           Talidex_Tab 25mg      1   

            cost  
0   673365.11672  
1   696800.10164  
2   507441.91000  
3   848094.45000  
4   233409.60000  
5   211325.73693  
6   372638.81023  
7   332084.50744  
8   241424.19000  
9   353998.12394  
10  333567.36780  
11   95119.15000  
12  288139.51379  
13  292060.82786  
14   85355.04000  
15   25328.42536  
16   40154.81128  
17   72533.70000  
18   65745.18000  
19   54878.66000  
20   21229.71000  
21   60015.56493  
22   52742.40000  
23   21712.69000  
24    4253.76445  
25     104.57000

**Individual Patient Supply** denotes medications which are only prescribed for a select number of patients

<a id='f2-4'></a>
#### Imported (avail_restrict 4)

In [8]:
sql = '''
SELECT id, nm, descr, bnf_code 
FROM `ebmdatalab.dmd.amp` test_DF
WHERE avail_restrict = 4
'''
avail_rest_4_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_rest_4.csv'))
avail_rest_4_DF

id  \
0    29175011000001104   
1    29193611000001104   
2    31584411000001103   
3    31736011000001109   
4    32641911000001105   
..                 ...   
654  19390111000001109   
655  19390511000001100   
656  19390811000001102   
657  19391111000001103   
658  10826611000001106   

                                                                    nm  \
0                                                 Iperten 10mg tablets   
1                                                   Proctolog ointment   
2                                            Dronabinol 2.5mg capsules   
3                                               Lumitene 30mg capsules   
4                                          Triapten Antiviral 2% cream   
..                                                                 ...   
654  Abnoba Viscum Q 200micrograms/1ml solution for injection ampoules   
655           Abnoba Viscum Q 20mg/1ml solution for injection ampoules   
656   Abnoba Viscum Q 20micrograms/1ml solution for injection ampoules   
657            Abnoba Viscum Q 2mg/1ml solution for injection ampoules   
658                                           ORA-Plus oral suspension   

                                                                                      descr  \
0                                                  Iperten 10mg tablets (Imported (France))   
1                                                     Proctolog ointment (Imported (Spain))   
2                                      Dronabinol 2.5mg capsules (Imported (United States))   
3                                         Lumitene 30mg capsules (Imported (United States))   
4                                          Triapten Antiviral 2% cream (Imported (Germany))   
..                                                                                      ...   
654  Abnoba Viscum Q 200micrograms/1ml solution for injection ampoules (Imported (Germany))   
655           Abnoba Viscum Q 20mg/1ml solution for injection ampoules (Imported (Germany))   
656   Abnoba Viscum Q 20micrograms/1ml solution for injection ampoules (Imported (Germany))   
657            Abnoba Viscum Q 2mg/1ml solution for injection ampoules (Imported (Germany))   
658                                     ORA-Plus oral suspension (Imported (United States))   

            bnf_code  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
..               ...  
654  190203000BBAAA0  
655  190203000BBAAA0  
656  190203000BBAAA0  
657  190203000BBAAA0  
658  190700000BBCQA0  

[659 rows x 4 columns]

In [9]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    SELECT DISTINCT bnf_code 
    FROM `ebmdatalab.dmd.amp` 
    WHERE avail_restrict = 4
)
GROUP BY bnf_code, bnf_name
ORDER BY items DESC
'''
avail_rest_4_usage_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_restrict_4_usage.csv'))

with pd.option_context('display.max_rows', 100):
    with pd.option_context('display.float_format', '{:,.2f}'.format):
        display(avail_rest_4_usage_DF.head(100))

Downloading: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|


bnf_code                                                 bnf_name  \
0   0602010V0AABZBZ                Levothyroxine sodium 100microgram tablets   
1   0602010V0AABXBX                 Levothyroxine sodium 50microgram tablets   
2   0602010V0AABWBW                 Levothyroxine sodium 25microgram tablets   
3   0602010V0AABZBZ                                Levothyrox Sod_Tab 100mcg   
4   0602010V0AABXBX                                 Levothyrox Sod_Tab 50mcg   
5   0602010V0AABWBW                                 Levothyrox Sod_Tab 25mcg   
6   1003020AAAAABAB                                   Capsaicin 0.025% cream   
7   1003020AAAAAAAA                                   Capsaicin 0.075% cream   
8   0301020I0AAACAC   Ipratropium bromide 500micrograms/2ml neb liq ud vials   
9   1003020AAAAABAB                                     Capsaicin_Crm 0.025%   
10  0604011K0AAABAB                           Estradiol valerate 2mg tablets   
11  0401010ADAAAEAE                                   Melatonin 2mg capsules   
12  0301020I0AAACAC                     Ipratrop Brom_Inh Soln 500mcg/2ml Ud   
13  0301020I0AAAFAF   Ipratropium bromide 250micrograms/1ml neb liq ud vials   
14  0404000M0AAANAN            Methylphenidate 36mg modified-release tablets   
15  0401010ADAABRBR                                   Melatonin 3mg capsules   
16  1003020AAAAAAAA                                     Capsaicin_Crm 0.075%   
17  0404000M0AAAMAM            Methylphenidate 18mg modified-release tablets   
18  1307000V0AAAAAA                         Imiquimod 5% cream 250mg sachets   
19  0404000M0AAANAN                         Methylphenidate HCl_Tab 36mg M/R   
20  1310011Y0AAAAAA                             Sulfadiazine silver 1% cream   
21  0404000M0AAATAT            Methylphenidate 27mg modified-release tablets   
22  0301020I0AAAFAF                     Ipratrop Brom_Inh Soln 250mcg/1ml Ud   
23  0203020F0AAABAB                              Disopyramide 100mg capsules   
24  0401010ADAABRBR                                        Melatonin_Cap 3mg   
25  0401010ADAAAEAE                                        Melatonin_Cap 2mg   
26  0404000M0AAAMAM                         Methylphenidate HCl_Tab 18mg M/R   
27  0401010ADAABSBS                                   Melatonin 5mg capsules   
28  0604011K0AAABAB                                    Estradiol_Val Tab 2mg   
29  0102000J0AAAAAA                     Dicycloverine HCl_Oral Soln 10mg/5ml   
30  0404000M0AAARAR            Methylphenidate 54mg modified-release tablets   
31  1307000V0AAAAAA                              Imiquimod_Crm 5% Sach 250mg   
32  1310011Y0AAAAAA                               Silver Sulfadiazine_Crm 1%   
33  091200000BFWRFT                                             Dialyvit_Tab   
34  0404000M0AAATAT                         Methylphenidate HCl_Tab 27mg M/R   
35  0906040G0AAEAEA                        Colecalciferol 5,000unit capsules   
36  0102000J0AAAAAA                     Dicycloverine 10mg/5ml oral solution   
37  0203020F0AAABAB                                   Disopyramide_Cap 100mg   
38  0203020G0AAACAC              Disopyramide 250mg modified-release tablets   
39  190203000BBAAA0                   Homoeopathic Preparation - Proprietary   
40  0401010ADAABSBS                                        Melatonin_Cap 5mg   
41  0401010ADAAAKAK                                  Melatonin 10mg capsules   
42  190203000BBAAA0                                  Homeopathic preparation   
43  091200000BFWRFT                                         Dialyvit tablets   
44  0401010ADAAAJAJ                  Melatonin 3mg modified-release capsules   
45  0502050B0AAACAC                               Griseofulvin 500mg tablets   
46  0401010ADAAAIAI                                    Melatonin 1mg tablets   
47  0502050B0AAACAC                                   Griseofulvin_Tab 500mg   
48  0502050B0AAABAB                               Griseofulvin 125mg tablets   
49  0203020

**Imported** denotes imported medications. The 'descr' field details Imported and the country imported from. However the BNF code isn't unique to imported products - for example 0602010V0AABZBZ	Levothyroxine sodium 100microgram tablets - seems to include standard UK product and imported products.

<a id='f2-5'></a>
#### Clinical Trial (avail_restrict 5)

In [10]:
sql = '''
SELECT id, nm, descr, bnf_code 
FROM `ebmdatalab.dmd.amp` 
WHERE avail_restrict = 5
'''
avail_rest_5_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_rest_5.csv'))
avail_rest_5_DF

id  \
0  41823211000001105   

                                                                                              nm  \
0  COVID-19 Vaccine Moderna (mRNA-1283.222) 10micrograms/0.2ml dose solution for injection vials   

                                                                                                          descr  \
0  COVID-19 Vaccine Moderna (mRNA-1283.222) 10micrograms/0.2ml dose solution for injection vials (Moderna, Inc)   

          bnf_code  
0  1404000ATBRADBJ

In [11]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    SELECT DISTINCT bnf_code 
    FROM `ebmdatalab.dmd.amp` 
    WHERE avail_restrict = 5
)
GROUP BY bnf_code, bnf_name
ORDER BY items DESC
'''
avail_rest_5_usage_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_restrict_5_usage.csv'))

avail_rest_5_usage_DF

Empty DataFrame
Columns: [bnf_code, bnf_name, items, cost]
Index: []

**Clinical Trial** denotes a medication used for clinical trial. Only one product, no prescribing.

<a id='f2-6'></a>
#### Specials (avail_restrict 6)

In [12]:
sql = '''
SELECT id, nm, descr, bnf_code 
FROM `ebmdatalab.dmd.amp` 
WHERE avail_restrict = 6
'''
avail_rest_6_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_rest_6.csv'))
avail_rest_6_DF

id                                                   nm  \
0     19980211000001105  Potassium ascorbate 10% eye drops preservative free   
1     20399911000001104                 Coal tar 30% in Emulsifying ointment   
2     25192611000001104                   Iodine 1% in Boric acid ear powder   
3     25407511000001104                Hydrochloric acid 18% dental solution   
4     25566311000001101                 Fentanyl 24.75mg/990ml infusion bags   
...                 ...                                                  ...   
7276  32284311000001104               Cade oil 12.5% in Emulsifying ointment   
7277  24551811000001101                                         Ginger syrup   
7278  18159811000001109                                    Eosin 1% solution   
7279  17656311000001103                                    Eosin 2% solution   
7280  23609411000001101                               Sodium benzoate powder   

                                                                    descr  \
0     Potassium ascorbate 10% eye drops preservative free (Special Order)   
1                    Coal tar 30% in Emulsifying ointment (Special Order)   
2                      Iodine 1% in Boric acid ear powder (Special Order)   
3                   Hydrochloric acid 18% dental solution (Special Order)   
4                    Fentanyl 24.75mg/990ml infusion bags (Special Order)   
...                                                                   ...   
7276               Cade oil 12.5% in Emulsifying ointment (Special Order)   
7277                                         Ginger syrup (Special Order)   
7278                                    Eosin 1% solution (Special Order)   
7279                        Eosin 2% solution (Drug Tariff Special Order)   
7280                               Sodium benzoate powder (Special Order)   

             bnf_code  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
...               ...  
7276  190605000AACECE  
7277  190607000AAAGAG  
7278  190800000AAAPAP  
7279  190800000AABTBT  
7280  190900000AAASAS  

[7281 rows x 4 columns]

In [13]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    SELECT DISTINCT bnf_code 
    FROM `ebmdatalab.dmd.amp` 
    WHERE avail_restrict = 6
)
GROUP BY bnf_code, bnf_name
ORDER BY items DESC
'''
avail_rest_6_usage_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_restrict_6_usage.csv'))

with pd.option_context('display.max_rows', 100):
    with pd.option_context('display.float_format', '{:,.2f}'.format):
        display(avail_rest_6_usage_DF.head(100))

bnf_code  \
0   0410030C0AAAFAF   
1   0410030C0AAAAAA   
2   0410030C0AAAFAF   
3   0410030C0AAAAAA   
4   0901020D0AAAHAH   
5   0906040G0AAELEL   
6   0407010Q0AAAAAA   
7   0401010ADAACZCZ   
8   0407010Q0AAAAAA   
9   0906040G0AACRCR   
10  0601021H0AAABAB   
11  0401010ADAADADA   
12  0104020H0AAAAAA   
13  0601021H0AAABAB   
14  0103050P0AABLBL   
15  0409020N0AAAMAM   
16  0303020Z0AAABAB   
17  0107010AAAAAJAJ   
18  0202010V0AAANAN   
19  0902011U0AAAHAH   
20  0103030S0AAAAAA   
21  0103050P0AABNBN   
22  0103030S0AAAAAA   
23  0906040G0AABTBT   
24  0408020V0AAALAL   
25  0401010ADAABYBY   
26  0107010AAAAAKAK   
27  0304010J0AAAEAE   
28  0403030Q0AAAQAQ   
29  0906040G0AACQCQ   
30  0103050P0AABMBM   
31  1108010K0AACFCF   
32  0403010F0AAARAR   
33  0906040G0AACYCY   
34  0501120Q0AAAAAA   
35  0601021H0AAAAAA   
36  0407010H0AAA5A5   
37  0906040G0AAACAC   
38  0906040G0AADXDX   
39  0601021H0AAAAAA   
40  0106020C0AAAFAF   
41  0906040G0AACNCN   
42  0906040G0AAEAEA   
43  0603020J0AAAXAX   
44  190203000BBAAA0   
45  0202030S0AAEBEB   
46  0902011U0AAAHAH   
47  0906040G0AABFBF   
48  0102000L0AAADAD   
49  0401010ADAAAKAK   
50  0906040G0AADWDW   
51  0906060Q0AAACAC   
52  0209000C0AAAJAJ   
53  0704020J0AAAZAZ   
54  0906040G0AAFGFG   
55  0104020H0AAAAAA   
56  0905013G0AAA4A4   
57  190203000BBAAA0   
58  0401010ADAADEDE   
59  0906040N0AAEWEW   
60  0107010AAAAAJAJ   
61  0906040G0AACPCP   
62  0906025P0AABMBM   
63  0103050P0AABLBL   
64  0103050L0AAAZAZ   
65  0905013G0AAA2A2   
66  1001050A0AAADAD   
67  0101010N0AAABAB   
68  0906025P0AABFBF   
69  1108010K0AAAAAA   
70  0906025P0AAAUAU   
71  0103050L0AAAYAY   
72  1303000AAAAAFAF   
73  0504040M0AAAAAA   
74  0906022K0AAAPAP   
75  0408010N0AACLCL   
76  0401010ADAACDCD   
77  0401010ADAAAJAJ   
78  0901020D0AAAHAH   
79  0607020C0AAAAAA   
80  0905013G0AABMBM   
81  0406000J0AAABAB   
82  0905013F0AAADAD   
83  0905013F0AAABAB   
84  0402020L0AAACAC   
85  0905013G0AADFDF   
86  0408010Z0AAACAC   
87  0103050L0AAAXAX   
88  0905013G0AAA2A2   
89  0202030S0AAEAEA   
90  0905013F0AAACAC   
91  0403030Q0AAARAR   
92  0204000H0AABEBE   
93  0202010D0AAAUAU   
94  0103010T0AABQBQ   
95  0401010ADAACECE   
96  0402010Q0AAACAC   
97  0101010I0AAACAC   
98  0103010T0AABRBR   
99  0401010ADAACFCF   

                                                      bnf_name    items  \
0                    Methadone 1mg/ml oral solution sugar free  9126099   
1                               Methadone 1mg/ml oral solution  4742757   
2                                Methadone HCl_Mix 1mg/1ml S/F  2017025   
3                                    Methadone HCl_Mix 1mg/1ml  1869833   
4                  Cyanocobalamin 1mg modified-release tablets  1216501   
5                               Colecalciferol 400unit tablets   693619   
6                                 Co-Proxamol_Tab 32.5mg/325mg   551243   
7                    Melatonin 1mg/ml oral solution sugar free   484045   
8                             Co-proxamol 32.5mg/325mg tablets   350255   
9                            Colecalciferol 20,000unit tablets   297544   
10                                       Glibenclamide_Tab 5mg   296823   
11                       Melatonin 5mg/5ml oral solution (old)   244463   
12                               Co-Phenotrope_Tab 2.5mg/25mcg   241455   
13                                   Glibenclamide 5mg tablets   170535   
14                         Omeprazole 10mg/5ml oral suspension   156019   
15                                   Orphenadrine 50mg tablets   146236   
16                                    Zafirlukast 20mg tablets   145426   
17                                          Diltiazem 2% cream   137236   
18                                    Metolazone 2.5mg tablets   131639   
19                    Pot Chlor Slow_Tab 600mg (Pot 8mmol) M/R   128876   
20                                           Sucralfate_Tab 1g   127610   
21                          Omeprazol

<a id='f2-10'></a>
#### For Patient Safety Consideration (avail_restrict 10)

In [14]:
sql = '''
SELECT id, nm, descr, bnf_code 
FROM `ebmdatalab.dmd.amp` 
WHERE avail_restrict = 10
'''
avail_rest_10_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_rest_10.csv'))
avail_rest_10_DF

id  \
0  26345611000001105   
1  26346211000001102   
2  29718111000001107   
3  22235811000001100   

                                                         nm  \
0  Phenol 90% cutaneous solution - should NOT be prescribed   
1  Phenol 95% cutaneous solution - should NOT be prescribed   
2  Phenol 80% cutaneous solution - should NOT be prescribed   
3               Phenol liquefied - should NOT be prescribed   

                                                                      descr  \
0  Phenol 90% cutaneous solution - should NOT be prescribed (Special Order)   
1  Phenol 95% cutaneous solution - should NOT be prescribed (Special Order)   
2  Phenol 80% cutaneous solution - should NOT be prescribed (Special Order)   
3               Phenol liquefied - should NOT be prescribed (Special Order)   

          bnf_code  
0              NaN  
1              NaN  
2              NaN  
3  1311050P0AAABAB

In [15]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    SELECT DISTINCT bnf_code 
    FROM `ebmdatalab.dmd.amp` 
    WHERE avail_restrict = 10
)
GROUP BY bnf_code, bnf_name
ORDER BY items DESC
'''
avail_rest_10_usage_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','avail_restrict_4_usage.csv'))

avail_rest_10_usage_DF

Downloading: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|


bnf_code           bnf_name  items        cost
0  1311050P0AAABAB   Phenol liquefied     78  4465.70855
1  1311050P0AAABAB  Phenol Liquefied_     69  2016.57000

**For Patient Safety Consideration** denotes medication which appear not to be recommended for prescribing (all denote "should NOT be prescribed").

### Summary of avail_restrict  


| cd  | descr                             | row_count | comments                       |
|-----|-----------------------------------|-----------|--------------------------------|
| 1   | None                              | 89895     | No restrictions on availability     |
| 2   | Restricted Availability           | 56        | medications which are only prescribed from specialist services e.g. clozapine           |
| 3   | Individual Patient Supply         | 24        | Only prescribed for a select number of patients e.g. sodium amytal, BioBag   |
| 4   | Imported                          | 659       | Sourced from another country   |
| 5   | Clinical Trial                    | 1         | For research purposes          |
| 6   | Special                           | 7281      | Specials products     |
| 7   | Extemp                            | 0         | No items currently listed here|
| 9   | Not available                     | 61299     | Currently unavailable products          |
| 10  | For Patient Safety Consideration  | 4         | Products limited for safety reasons     |


<a id='#f3'></a>
### Focus on specials
Firstly look at how many specials are denoted by BNF code that only signifies a special.


In [16]:
sql = '''
SELECT 
    bnf_code,
    SUM(CASE WHEN avail_restrict = 6 THEN 1 ELSE 0 END) AS special_count,
    SUM(CASE WHEN avail_restrict != 6 THEN 1 ELSE 0 END) AS non_special_count
FROM 
    `ebmdatalab.dmd.amp`
WHERE
    bnf_code IS NOT NULL
GROUP BY 
    bnf_code
HAVING 
    special_count > 0
ORDER BY 
    special_count DESC, non_special_count DESC;
'''
spec_non_spec_count_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','spec_non_spec_count.csv'))
spec_non_spec_count_DF

bnf_code  special_count  non_special_count
0     190203000BBAAA0             46                397
1     1305020C0AAFPFP              6                  2
2     1307000M0AABVBV              4                  1
3     1305020C0AAAVAV              4                  0
4     1305020C0AABSBS              4                  0
...               ...            ...                ...
4432  190601000AABEBE              1                  0
4433  190605000AACDCD              1                  0
4434  190605000AACECE              1                  0
4435  190800000AAAPAP              1                  0
4436  190800000AABTBT              1                  0

[4437 rows x 3 columns]

In [17]:
sql = '''
SELECT 
    amp.bnf_code,
    SUM(CASE WHEN amp.avail_restrict = 6 THEN 1 ELSE 0 END) AS special_count,
    SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9) THEN 1 ELSE 0 END) AS non_special_count,
    STRING_AGG(DISTINCT CASE WHEN amp.avail_restrict = 6 THEN amp.descr ELSE NULL END, ', ') AS special_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9) THEN amp.descr ELSE NULL END, ', ') AS non_special_names,
    STRING_AGG(DISTINCT vmp.nm, ', ') AS vpm_names
FROM 
    `ebmdatalab.dmd.amp` AS amp
LEFT JOIN 
    `ebmdatalab.dmd.vmp` AS vmp
ON 
    amp.bnf_code = vmp.bnf_code
WHERE
    amp.bnf_code IS NOT NULL
GROUP BY 
    amp.bnf_code
HAVING 
    special_count > 0
ORDER BY 
    non_special_count DESC;
'''
spec_non_spec_count_named_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','spec_non_spec_count_named.csv'))

with pd.option_context('display.max_rows', 100):
    display(spec_non_spec_count_named_DF.head(100))


bnf_code  special_count  non_special_count  \
0   190203000BBAAA0          16238              37771   
1   0401010ADAACZCZ              1                  7   
2   0410030C0AAAAAA              1                  5   
3   0410030C0AAAFAF              1                  5   
4   0407010H0AAA5A5              1                  4   
5   0906012B0AAACAC              1                  2   
6   0906025P0AABFBF              1                  2   
7   0402010AFAAAAAA              1                  2   
8   0702020T0AAAFAF              1                  1   
9   090501300AABIBI              1                  1   
10  0905013F0AAABAB              1                  1   
11  090504100AABBBB              1                  1   
12  0906022K0AAAPAP              1                  1   
13  0913445A0AAAGAG              1                  1   
14  0203020U0AAAGAG              1                  1   
15  0401010ADAAACAC              1                  1   
16  0401010ADAAAHAH              1                  1   
17  0401010ADAAAJAJ              1                  1   
18  0401010ADAAAKAK              1                  1   
19  0401010ADAAAQAQ              1                  1   
20  0401010ADAABIBI              1                  1   
21  0401010ADAABQBQ              1                  1   
22  0402010AHAAAAAA              1                  1   
23  0406000W0AAAMAM              1                  1   
24  0407020Y0AAADAD              1                  1   
25  0408020W0AAAHAH              1                  1   
26  0409020G0AAACAC              1                  1   
27  0501021E0AAADAD              1                  1   
28  0501021E0AAAEAE              1                  1   
29  0501030F0AAAIAI              1                  1   
30  0501080T0AAAAAA              1                  1   
31  0602010V0AAFWFW              1                  1   
32  0605020E0AAASAS              1                  1   
33  0801030L0AAABAB              1                  1   
34  0801050BUAAAAAA              1                  1   
35  0801050BUAAABAB              1                  1   
36  0803020L0AAAAAA              1                  1   
37  0803020L0AAAGAG              1                  1   
38  0906031C0AAAPAP              1                  1   
39  0906040G0AACHCH              1                  1   
40  0906040G0AAEAEA              1                  1   
41  1001040K0AAADAD              1                  1   
42  1001040P0AAACAC              1                  1   
43  1502010G0AAAUAU              1                  1   
44  0101010C0AAAHAH              1                  0   
45  0101010C0AAAMAM              1                  0   
46  0101010F0AAAUAU              1                  0   
47  0101010I0AAABAB              1                  0   
48  0101010I0AAACAC              1                  0   
49  0101010I0AAADAD              1                  0   
50  0101010I0AAAHAH              1                  0   
51  0101010I0AAAJAJ              1                  0   
52  0101010I0AAALAL              1                  0   
53  0101010I0AAAMAM              1                  0   
54  0101010I0AAATAT              1                  0   
55  0101010I0AAAUAU              1                  0   
56  0101010I0AAAVAV              1                  0   
57  0101010I0AAAXAX              1                  0   
58  0101010I0AAAYAY              1                  0   
59  0101010I0AABGBG              1                  0   
60  0101010I0AABIBI              1                  0   
61  0101010I0AABPBP              1                  0   
62  0101010I0AABQBQ              1                  0   
63  0101010I0AABRBR              1                  0   
64  0101010I0AABSBS              1                  0   
65  0101010I0AABTBT              1                  0   
66  0101010N0AAABAB              1                  0   
67  0101012B0AAALAL              1                  0   
68  0101012B0AAATAT              1                  0   
69  0101012B0AAAUAU  

BNF code 190203000BBAAA0 covers all homeopathic preperations. This is a mix of specials and none specials and cannot be differentiated. Prescribing of these products is minimal - £2,625 across England Nov '23—Oct '24. So will exclude this code (and detail this in any WIM).

Looking further at the list a number of products have 1 special product and 1 imported product listed under the same BNF code - without any none special/none imported product. Given that the imported products would be unlicensed in nature it seems reasonable to include prescribing under these BNF codes. But worth investigating further how imported products link to specials and none specials.

In [18]:
sql = '''
SELECT 
    amp.bnf_code,
    SUM(CASE WHEN (amp.avail_restrict = 6) THEN 1 ELSE 0 END) AS special_count,
    SUM(CASE WHEN (amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS import_count,
    SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count,
    STRING_AGG(DISTINCT CASE WHEN amp.avail_restrict = 6 THEN amp.descr ELSE NULL END, ', ') AS special_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict = 4) THEN amp.descr ELSE NULL END, ', ') AS import_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN amp.descr ELSE NULL END, ', ') AS non_special_names,
    STRING_AGG(DISTINCT vmp.nm, ', ') AS vpm_names
FROM 
    `ebmdatalab.dmd.amp` AS amp
LEFT JOIN 
    `ebmdatalab.dmd.vmp` AS vmp
ON 
    amp.bnf_code = vmp.bnf_code
WHERE
    amp.bnf_code IS NOT NULL
    AND
    amp.bnf_code != '190203000BBAAA0' --  Exclude homeopathic products as total mix of specials and none specials but minimal impact on cost ~£2600 across England over 12 months
GROUP BY 
    amp.bnf_code
HAVING 
    special_count > 0
    AND
    import_count >0
    AND
    non_special_non_import_count = 0
ORDER BY 
    import_count DESC;
'''
spec_imported_non_spec_count_named_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','spec_imported_non_spec_count_named.csv'))

with pd.option_context('display.max_rows', 100):
    display(spec_imported_non_spec_count_named_DF.head(100))

bnf_code  special_count  import_count  \
0   0402010AFAAAAAA              1             2   
1   0203020U0AAAGAG              1             1   
2   0401010ADAAACAC              1             1   
3   0401010ADAAAHAH              1             1   
4   0401010ADAAAJAJ              1             1   
5   0401010ADAAAKAK              1             1   
6   0401010ADAAAQAQ              1             1   
7   0401010ADAABIBI              1             1   
8   0401010ADAABQBQ              1             1   
9   0402010AHAAAAAA              1             1   
10  0406000W0AAAMAM              1             1   
11  0407020Y0AAADAD              1             1   
12  0408020W0AAAHAH              1             1   
13  0409020G0AAACAC              1             1   
14  0501021E0AAADAD              1             1   
15  0501021E0AAAEAE              1             1   
16  0501030F0AAAIAI              1             1   
17  0501080T0AAAAAA              1             1   
18  0602010V0AAFWFW              1             1   
19  0605020E0AAASAS              1             1   
20  0801030L0AAABAB              1             1   
21  0801050BUAAAAAA              1             1   
22  0801050BUAAABAB              1             1   
23  0803020L0AAAAAA              1             1   
24  0803020L0AAAGAG              1             1   
25  0906031C0AAAPAP              1             1   
26  0906040G0AACHCH              1             1   
27  0906040G0AAEAEA              1             1   
28  1001040K0AAADAD              1             1   
29  1001040P0AAACAC              1             1   
30  1502010G0AAAUAU              1             1   

    non_special_non_import_count  \
0                              0   
1                              0   
2                              0   
3                              0   
4                              0   
5                              0   
6                              0   
7                              0   
8                              0   
9                              0   
10                             0   
11                             0   
12                             0   
13                             0   
14                             0   
15                             0   
16                             0   
17                             0   
18                             0   
19                             0   
20                             0   
21                             0   
22                             0   
23                             0   
24                             0   
25                             0   
26                             0   
27                             0   
28                             0   
29                             0   
30                             0   

                                                                                special_names  \
0                                       Melperone hydrochloride 100mg tablets (Special Order)   
1                                             Quinidine sulfate 200mg tablets (Special Order)   
2                                             Melatonin 500microgram capsules (Special Order)   
3                             Melatonin 2.5mg capsules (Part VIIID Drug Tariff Special Order)   
4              Melatonin 3mg modified-release capsules (Part VIIID Drug Tariff Special Order)   
5                                                     Melatonin 10mg capsules (Special Order)   
6                                      Melatonin 3mg modified-release tablets (Special Order)   
7                                           Melatonin 3mg lozenges sugar free (Special Order)   
8                                                      Melatonin 1mg capsules (Special Order)   
9                                                     Clotiapine 40mg tablets (Special Order)   
10                                              Ketamine hydrochloride powder (Special Order)   
11                        Nal

All BNF codes with only AMPs which have a special or imported status would be reasonable to include in the measure.

We should also look at identifying which BNF codes have an imported product and none special product.

In [19]:
sql = '''
SELECT 
    amp.bnf_code,
    SUM(CASE WHEN (amp.avail_restrict = 6) THEN 1 ELSE 0 END) AS special_count,
    SUM(CASE WHEN (amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS import_count,
    SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count,
    STRING_AGG(DISTINCT CASE WHEN amp.avail_restrict = 6 THEN amp.descr ELSE NULL END, ', ') AS special_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict = 4) THEN amp.descr ELSE NULL END, ', ') AS import_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN amp.descr ELSE NULL END, ', ') AS non_special_names,
    STRING_AGG(DISTINCT vmp.nm, ', ') AS vpm_names
FROM 
    `ebmdatalab.dmd.amp` AS amp
LEFT JOIN 
    `ebmdatalab.dmd.vmp` AS vmp
ON 
    amp.bnf_code = vmp.bnf_code
WHERE
    amp.bnf_code IS NOT NULL
    AND
    amp.bnf_code != '190203000BBAAA0' --  Exclude homeopathic products as total mix of specials and none specials but minimal impact on cost ~£2600 across England over 12 months
GROUP BY 
    amp.bnf_code
HAVING 
    import_count >0
    AND
    non_special_non_import_count > 0
ORDER BY 
    import_count DESC;
'''
imported_non_spec_count_named_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','imported_non_spec_count_named.csv'))

with pd.option_context('display.max_rows', 100):
    display(imported_non_spec_count_named_DF.head(100))

bnf_code  special_count  import_count  \
0   0401010ADAABRBR              0             2   
1   0102000J0AAAAAA              0             1   
2   0203020P0AAANAN              0             1   
3   0301020I0AAACAC              0             1   
4   0301020I0AAAFAF              0             1   
5   0401010ADAAAEAE              0             1   
6   0401010ADAABSBS              0             1   
7   0502050B0AAABAB              0             1   
8   0502050B0AAACAC              0             1   
9   0602010V0AABWBW              0             1   
10  0602010V0AABXBX              0             1   
11  0602010V0AABZBZ              0             1   
12  0604011K0AAABAB              0             1   
13  0605010P0AAAHAH              0             1   

    non_special_non_import_count  special_names  \
0                              3            NaN   
1                              6            NaN   
2                              3            NaN   
3                              6            NaN   
4                              6            NaN   
5                              4            NaN   
6                              3            NaN   
7                              3            NaN   
8                              3            NaN   
9                             11            NaN   
10                            11            NaN   
11                            12            NaN   
12                             2            NaN   
13                             1            NaN   

                                                                                      import_names  \
0   Melatonin 3mg capsules (Imported (United States)), Melatonin 3mg capsules (Imported (Denmark))   
1                                  Dicycloverine 10mg/5ml oral solution (Imported (United States))   
2            Mexiletine hydrochloride 200mg (Mexiletine 167mg) capsules (Imported (United States))   
3                Ipratropium bromide 500micrograms/2ml nebuliser liquid unit dose vials (Imported)   
4                Ipratropium bromide 250micrograms/1ml nebuliser liquid unit dose vials (Imported)   
5                                                      Melatonin 2mg capsules (Imported (Denmark))   
6                                                Melatonin 5mg capsules (Imported (United States))   
7                                            Griseofulvin 125mg tablets (Imported (United States))   
8                                            Griseofulvin 500mg tablets (Imported (United States))   
9                       Levothyroxine sodium 25microgram tablets lactose free (Imported (Germany))   
10                      Levothyroxine sodium 50microgram tablets lactose free (Imported (Germany))   
11                     Levothyroxine sodium 100microgram tablets lactose free (Imported (Germany))   
12                                             Estradiol valerate 2mg tablets (Imported (Germany))   
13                     Gonadorelin 100microgram powder for solution for injection vials (Imported)   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              non_special_names  \
0                                                                                                 

Only a small number of products. Add spend column to understand significance of the spending against that BNF code.

In [20]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    WITH aggregated_data AS (
        SELECT 
            amp.bnf_code,
            SUM(CASE WHEN (amp.avail_restrict = 6) THEN 1 ELSE 0 END) AS special_count,
            SUM(CASE WHEN (amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS import_count,
            SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count
        FROM 
            `ebmdatalab.dmd.amp` AS amp
        WHERE
            amp.bnf_code IS NOT NULL
            AND amp.bnf_code != '190203000BBAAA0'
        GROUP BY 
            amp.bnf_code
    )
    SELECT DISTINCT bnf_code
    FROM aggregated_data
    WHERE import_count > 0
    AND non_special_non_import_count > 0
)
AND month >='2023-11-01'
AND month < '2024-11-01'
GROUP BY bnf_code, bnf_name
ORDER BY cost DESC
'''
mixed_imported_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','mixed_imported.csv'))
with pd.option_context('display.max_rows', 100):
    with pd.option_context('display.float_format', '{:,.2f}'.format):
        display(mixed_imported_DF.head(100))

bnf_code  \
0   0602010V0AABZBZ   
1   0602010V0AABWBW   
2   0602010V0AABXBX   
3   0401010ADAAAEAE   
4   0401010ADAABRBR   
5   0203020P0AAANAN   
6   0401010ADAABSBS   
7   0102000J0AAAAAA   
8   0301020I0AAACAC   
9   0301020I0AAAFAF   
10  0604011K0AAABAB   
11  0502050B0AAACAC   
12  0502050B0AAABAB   
13  0605010P0AAAHAH   

                                                      bnf_name     items  \
0                    Levothyroxine sodium 100microgram tablets  12020168   
1                     Levothyroxine sodium 25microgram tablets   9947139   
2                     Levothyroxine sodium 50microgram tablets  10253041   
3                                       Melatonin 2mg capsules     33438   
4                                       Melatonin 3mg capsules     17362   
5   Mexiletine hydrochloride 200mg (Mexiletine 167mg) capsules       186   
6                                       Melatonin 5mg capsules      4588   
7                         Dicycloverine 10mg/5ml oral solution      1239   
8       Ipratropium bromide 500micrograms/2ml neb liq ud vials     42022   
9       Ipratropium bromide 250micrograms/1ml neb liq ud vials     16225   
10                              Estradiol valerate 2mg tablets     32228   
11                                  Griseofulvin 500mg tablets       387   
12                                  Griseofulvin 125mg tablets       156   
13             Gonadorelin 100microgram pdr for soln inj vials        25   

           cost  
0  9,503,331.59  
1  8,730,341.78  
2  7,767,556.64  
3  2,683,691.49  
4  1,444,806.38  
5    810,406.93  
6    651,839.03  
7    602,816.85  
8    597,169.05  
9    298,477.10  
10   242,122.28  
11    18,713.37  
12    12,266.84  
13     1,763.03

For many of these it is likely the vast majority of the spend is on the UK licensed product - with only a small amount for specific patient cases - e.g. levothyroxine lactose free imported.
We may just need to exclude data for these products from any measure as we cannot reliably quantify the usage of imported vs non-imported from EPD.

We also need to consider mixed specials/imports products with UK licensed products

In [21]:
sql = '''
SELECT 
    amp.bnf_code,
    SUM(CASE WHEN (amp.avail_restrict = 6) THEN 1 ELSE 0 END) AS special_count,
    SUM(CASE WHEN (amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS import_count,
    SUM(CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS special_import_count,
    SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN amp.descr ELSE NULL END, ', ') AS special_import_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN amp.descr ELSE NULL END, ', ') AS non_special_import_names,
    STRING_AGG(DISTINCT vmp.nm, ', ') AS vpm_names
FROM 
    `ebmdatalab.dmd.amp` AS amp
LEFT JOIN 
    `ebmdatalab.dmd.vmp` AS vmp
ON 
    amp.bnf_code = vmp.bnf_code
WHERE
    amp.bnf_code IS NOT NULL
    AND
    amp.bnf_code != '190203000BBAAA0' --  Exclude homeopathic products as total mix of specials and none specials but minimal impact on cost ~£2600 across England over 12 months
GROUP BY 
    amp.bnf_code
HAVING
    special_count > 0
    AND
    special_import_count >0
    AND
    non_special_non_import_count > 0
ORDER BY 
    special_import_count DESC;
'''
mixed_count_named_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','mixed_count_named.csv'))

with pd.option_context('display.max_rows', 100):
    display(mixed_count_named_DF.head(100))

bnf_code  special_count  import_count  special_import_count  \
0   0401010ADAACZCZ              1             0                     1   
1   0407010H0AAA5A5              1             0                     1   
2   0410030C0AAAAAA              1             0                     1   
3   0410030C0AAAFAF              1             0                     1   
4   0702020T0AAAFAF              1             0                     1   
5   090501300AABIBI              1             0                     1   
6   0905013F0AAABAB              1             0                     1   
7   090504100AABBBB              1             0                     1   
8   0906012B0AAACAC              1             0                     1   
9   0906022K0AAAPAP              1             0                     1   
10  0906025P0AABFBF              1             0                     1   
11  0913445A0AAAGAG              1             0                     1   

    non_special_non_import_count  \
0                              7   
1                              4   
2                              5   
3                              5   
4                              1   
5                              1   
6                              1   
7                              1   
8                              2   
9                              1   
10                             2   
11                             1   

                                                                               special_import_names  \
0                            Melatonin 1mg/ml oral solution alcohol free sugar free (Special Order)   
1   Paracetamol 500mg/5ml oral suspension (Royal Preston Hospital mucilage formula) (Special Order)   
2                                        Methadone 1mg/ml oral solution colour free (Special Order)   
3                             Methadone 1mg/ml oral solution colour free sugar free (Special Order)   
4                                                    Nystatin 100,000unit pessaries (Special Order)   
5                                                   Magnesium malate 250mg capsules (Special Order)   
6                                                  Magnesium citrate 100mg capsules (Special Order)   
7                                                        Zinc citrate 50mg capsules (Special Order)   
8                                                        Betacarotene 15mg capsules (Special Order)   
9                                                        Nicotinamide 250mg tablets (Special Order)   
10                                                         Riboflavin 50mg capsules (Special Order)   
11                                                   L-Phenylalanine 500mg capsules (Special Order)   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       non_special_import_names  \
0   Melatonin 1mg/ml oral solution sugar free (Colonis Pharma Ltd), Melatonin 1mg/ml oral solution sugar free (A A H Pharmaceuticals Ltd), Melatonin 1mg/ml oral solution sugar free (Alliance Healthcare (Distribution) Ltd), Melatonin 1mg/ml oral solution sugar free (Consilient Health Ltd), Melatonin 1mg/ml oral solution sugar free (Aspire Pharma Ltd), Melatonin 1mg/ml oral solution sugar free (Alissa Healthcare Research Ltd), Melatonin 1mg/ml oral solution sugar free (Thame Laboratories Ltd)   
1                                                                                                                                                                        Paracetamol 500mg/5ml oral suspension sugar free (Rosemont Pharmaceutica

Only a small number of products. Add spend column to understand significance of the spending against that BNF code.

In [22]:
sql = '''
SELECT bnf_code, 
       bnf_name, 
       SUM(items) AS items, 
       SUM(actual_cost) AS cost
FROM `ebmdatalab.hscic.normalised_prescribing`
WHERE bnf_code IN (
    WITH aggregated_data AS (
        SELECT 
            amp.bnf_code,
            SUM(CASE WHEN (amp.avail_restrict = 6) THEN 1 ELSE 0 END) AS special_count,
            SUM(CASE WHEN (amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS import_count,
            SUM(CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS special_import_count,
            SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count
        FROM 
            `ebmdatalab.dmd.amp` AS amp
        WHERE
            amp.bnf_code IS NOT NULL
            AND amp.bnf_code != '190203000BBAAA0'
        GROUP BY 
            amp.bnf_code
    )
    SELECT DISTINCT bnf_code
    FROM aggregated_data
    WHERE special_count > 0
    AND
    special_import_count >0
    AND
    non_special_non_import_count > 0
    
)
AND month >='2023-11-01'
AND month < '2024-11-01'
GROUP BY bnf_code, bnf_name
ORDER BY cost DESC
'''
mixed_specials_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','mixed_specials.csv'))
with pd.option_context('display.max_rows', 100):
    with pd.option_context('display.float_format', '{:,.2f}'.format):
        display(mixed_specials_DF.head(100))

bnf_code                                          bnf_name   items  \
0   0401010ADAACZCZ         Melatonin 1mg/ml oral solution sugar free  123236   
1   0410030C0AAAFAF         Methadone 1mg/ml oral solution sugar free  705950   
2   0410030C0AAAAAA                    Methadone 1mg/ml oral solution  214491   
3   0906025P0AABFBF                          Riboflavin 50mg capsules    4849   
4   0407010H0AAA5A5  Paracetamol 500mg/5ml oral suspension sugar free    6863   
5   0906022K0AAAPAP                        Nicotinamide 250mg tablets    2891   
6   0905013F0AAABAB                  Magnesium citrate 100mg capsules    3559   
7   0702020T0AAAFAF                    Nystatin 100,000unit pessaries     487   
8   090504100AABBBB                        Zinc citrate 50mg capsules    1037   
9   0906012B0AAACAC                        Betacarotene 15mg capsules      16   
10  090501300AABIBI                   Magnesium malate 250mg capsules     321   
11  0913445A0AAAGAG                    L-Phenylalanine 500mg capsules     171   

            cost  
0  13,883,526.23  
1   4,256,744.13  
2   1,312,512.73  
3     706,847.98  
4     653,555.52  
5     357,191.72  
6     228,914.70  
7     179,385.75  
8      11,938.72  
9       6,659.56  
10      6,370.09  
11      2,782.95

Again, for many of these it is likely the vast majority of the spend is on the UK licensed product

### Building final query
From the profiling above, for this measure we will look to include all BNF codes for special products or imported products except when they cannot be reliably differentiated from UK licensed products.
We want to include:
- all BNF codes where specials + import count > 0

Except:
- where non_special_non_import_count also >0

#### Excluded results

In [23]:
sql = '''
SELECT 
    amp.bnf_code,
    SUM(CASE WHEN (amp.avail_restrict = 6) THEN 1 ELSE 0 END) AS special_count,
    SUM(CASE WHEN (amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS import_count,
    SUM(CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS special_import_count,
    SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN amp.descr ELSE NULL END, ', ') AS special_import_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN amp.descr ELSE NULL END, ', ') AS non_special_import_names,
    STRING_AGG(DISTINCT vmp.nm, ', ') AS vpm_names
FROM 
    `ebmdatalab.dmd.amp` AS amp
LEFT JOIN 
    `ebmdatalab.dmd.vmp` AS vmp
ON 
    amp.bnf_code = vmp.bnf_code
WHERE
    amp.bnf_code IS NOT NULL
GROUP BY 
    amp.bnf_code
HAVING
    special_import_count > 0
    AND
    non_special_non_import_count > 0
ORDER BY 
    special_import_count DESC;
'''
excluded_list_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','excluded_list.csv'))

with pd.option_context('display.max_rows', 100):
    display(excluded_list_DF.head(100))

bnf_code  special_count  import_count  special_import_count  \
0   190203000BBAAA0          16238          5648                 21886   
1   0401010ADAABRBR              0             2                     2   
2   0102000J0AAAAAA              0             1                     1   
3   0203020P0AAANAN              0             1                     1   
4   0301020I0AAACAC              0             1                     1   
5   0301020I0AAAFAF              0             1                     1   
6   0401010ADAAAEAE              0             1                     1   
7   0401010ADAABSBS              0             1                     1   
8   0401010ADAACZCZ              1             0                     1   
9   0407010H0AAA5A5              1             0                     1   
10  0410030C0AAAAAA              1             0                     1   
11  0410030C0AAAFAF              1             0                     1   
12  0502050B0AAABAB              0             1                     1   
13  0502050B0AAACAC              0             1                     1   
14  0602010V0AABWBW              0             1                     1   
15  0602010V0AABXBX              0             1                     1   
16  0602010V0AABZBZ              0             1                     1   
17  0604011K0AAABAB              0             1                     1   
18  0605010P0AAAHAH              0             1                     1   
19  0702020T0AAAFAF              1             0                     1   
20  090501300AABIBI              1             0                     1   
21  0905013F0AAABAB              1             0                     1   
22  090504100AABBBB              1             0                     1   
23  0906012B0AAACAC              1             0                     1   
24  0906022K0AAAPAP              1             0                     1   
25  0906025P0AABFBF              1             0                     1   
26  0913445A0AAAGAG              1             0                     1   

    non_special_non_import_count  \
0                          32123   
1                              3   
2                              6   
3                              3   
4                              6   
5                              6   
6                              4   
7                              3   
8                              7   
9                              4   
10                             5   
11                             5   
12                             3   
13                             3   
14                            11   
15                            11   
16                            12   
17                             2   
18                             1   
19                             1   
20                             1   
21                             1   
22                             1   
23                             2   
24                             1   
25                             2   
26                             1   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

#### Included results

In [24]:
sql = '''
SELECT 
    amp.bnf_code,
    SUM(CASE WHEN (amp.avail_restrict = 6) THEN 1 ELSE 0 END) AS special_count,
    SUM(CASE WHEN (amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS import_count,
    SUM(CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS special_import_count,
    SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN amp.descr ELSE NULL END, ', ') AS special_import_names,
    STRING_AGG(DISTINCT CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN amp.descr ELSE NULL END, ', ') AS non_special_import_names,
    STRING_AGG(DISTINCT vmp.nm, ', ') AS vpm_names
FROM 
    `ebmdatalab.dmd.amp` AS amp
LEFT JOIN 
    `ebmdatalab.dmd.vmp` AS vmp
ON 
    amp.bnf_code = vmp.bnf_code
WHERE
    amp.bnf_code IS NOT NULL
GROUP BY 
    amp.bnf_code
HAVING
    special_import_count > 0
    AND
    non_special_non_import_count = 0
ORDER BY 
    special_import_count DESC;
'''
included_list_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','included_list.csv'))
included_list_DF

bnf_code  special_count  import_count  special_import_count  \
0     1305020C0AAFPFP             36             0                    36   
1     1305020C0AAAVAV             16             0                    16   
2     1305020C0AABSBS             16             0                    16   
3     1305020S0AACZCZ             16             0                    16   
4     1307000M0AABVBV             16             0                    16   
...               ...            ...           ...                   ...   
4706  190605000AACECE              1             0                     1   
4707  190607000AAAGAG              1             0                     1   
4708  190800000AAAPAP              1             0                     1   
4709  190800000AABTBT              1             0                     1   
4710  190900000AAASAS              1             0                     1   

      non_special_non_import_count  \
0                                0   
1                                0   
2                                0   
3                                0   
4                                0   
...                            ...   
4706                             0   
4707                             0   
4708                             0   
4709                             0   
4710                             0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                special_import_names  \
0     Coal tar solution 6% / Salicylic acid 2% in Emulsifying ointment (Drug Tariff Special Order), Coal tar solution 6% / Salicylic acid 2% in Unguentum M cream (Drug Tariff Special Order), Coal tar solution 6% / Salicylic acid 2% in ointment (Drug Tariff Special Order), Coal tar solution 6% / Salicylic acid 2% in White soft paraffin (Drug Tariff Special Order), Coal tar solution 6% / Salicylic acid 2% in Yellow soft paraffin (Drug Tariff Special Order), Coal tar solution 6% / Salicylic acid 2% in Epaderm ointment (Drug Tariff Special Order)   
1                                                                                                                                                                                                                                                          Coal tar 5% in Yellow soft paraffin (Part VIIID Drug Tariff Special Order), Coal tar 5% in Emulsifying ointment (Part VIIID Drug Tariff Special Order), Coal tar 5% in White soft paraffin (Part VIIID Drug Tariff Special Order), Coal tar 5% in Hydrous ointment (Part VIIID Drug Tariff Special Order)   
2                                                                                                                                                                                                                                                      Coal tar 10% in Yellow soft paraffin (Part VIIID Drug Tariff Special Order), Coal tar 10% in Emulsifying ointment (Part VIIID Drug Tariff Special Order), Coal tar 10% in White soft paraffin (Part VIIID Drug Tariff Special Order), Coal tar 10% in Hydrous ointment (Part VIIID Drug Tariff Special Order)   
3                                                                                                                                                                                                                                                                              Salicylic acid 5% in Emulsifying ointment (Drug Tariff Special Order), Salicylic acid 5% in White soft paraffin (Drug Tariff Special Order), Salicylic acid 5% in Ye

#### Included distinct BNF codes 

In [25]:
sql = '''
WITH aggregated_data AS (
    SELECT 
        amp.bnf_code,
        SUM(CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS special_import_count,
        SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count
    FROM 
        `ebmdatalab.dmd.amp` AS amp
    WHERE
        amp.bnf_code IS NOT NULL
    GROUP BY 
        amp.bnf_code
    )
SELECT DISTINCT bnf_code
FROM aggregated_data
WHERE special_import_count > 0
AND
non_special_non_import_count = 0;
'''
included_bnf_code_list_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','included_bnf_code_list.csv'))
included_bnf_code_list_DF

bnf_code
0     0101010I0BCAABI
1     0102000D0BHABAH
2     0102000R0BBAAAB
3     0103030S0BCAAAA
4     0103030S0BDAAAA
...               ...
4706  190605000AACECE
4707  190607000AAAGAG
4708  190800000AAAPAP
4709  190800000AABTBT
4710  190900000AAASAS

[4711 rows x 1 columns]

In [26]:
sql = '''
WITH aggregated_data AS (
    SELECT 
        amp.bnf_code,
        SUM(CASE WHEN (amp.avail_restrict = 6 OR amp.avail_restrict = 4) THEN 1 ELSE 0 END) AS special_import_count,
        SUM(CASE WHEN (amp.avail_restrict != 6 AND amp.avail_restrict != 9 AND amp.avail_restrict != 4) THEN 1 ELSE 0 END) AS non_special_non_import_count
    FROM 
        `ebmdatalab.dmd.amp` AS amp
    WHERE
        amp.bnf_code IS NOT NULL
    GROUP BY 
        amp.bnf_code
    )
SELECT DISTINCT bnf_code
FROM aggregated_data
WHERE special_import_count > 0
AND
non_special_non_import_count = 0;
'''
included_bnf_code_list_DF = bq.cached_read(sql, csv_path=os.path.join('../','data','included_bnf_code_list.csv'))
included_bnf_code_list_DF

bnf_code
0     0101010I0BCAABI
1     0102000D0BHABAH
2     0102000R0BBAAAB
3     0103030S0BCAAAA
4     0103030S0BDAAAA
...               ...
4706  190605000AACECE
4707  190607000AAAGAG
4708  190800000AAAPAP
4709  190800000AABTBT
4710  190900000AAASAS

[4711 rows x 1 columns]